hardfork-planning: initial dijkstra scope (#54)
IntersectMBO/product-website: docs/hardfork-planning/dijkstra.md

The distinction matters for what each phase can change. A new era ships a complete new ledger code package and can introduce new block structures, new serialization formats, new cryptography, and new Plutus versions. An intra-era hard fork bumps the protocol version within the existing era and can change anything gate-able on a protocol version check: activating consensus rules, enabling new Plutus primitives within an existing Plutus version, turning on features whose block-body structures were already defined by the era. Because Phase 1 ships the block body extensions that Linear Leios and Peras require, Phases 2 and 3 can activate those protocols without a new era.

| Phase | Mechanism | Era | Code Complete Target | Primary Activation |
|-------|-----------|-----|---------------------|--------------------|
| [Phase 1](#phase-1-dijkstra-hard-fork-protocol-version-12-october-2026) | New era (v12) | Dijkstra | October 2026 | Nested Transactions, block body extensions for Leios & Peras |
| [Phase 2](#phase-2-linear-leios-activation-intra-era-hard-fork-q1-2027) | Intra-era hard fork | Dijkstra | Q1 2027 | Ouroboros Linear Leios |

> | [Phase 3](#phase-3-peras-activation-intra-era-hard-fork-q2-2027) | Intra-era hard fork | Dijkstra | Q2 2027 | Ouroboros Peras |

Reviewer: Is the ecosystem ready for a hard-fork every 3 months? PV11 took more than 6 months after code freeze and past hard-forks with comparable scope to these took even longer.

#### Reference Script Pricing and Limits

The pricing and size limits for reference scripts have been in effect since Conway as hardcoded values. Dijkstra introduces them as proper protocol parameters. For the community to update them via governance actions, both the Constitution and the guardrails script must be updated to include bounds and rules for these new parameters.

#### Remove DRep Requirement for Reward Withdrawals ([CIP-181](https://cips.cardano.org/cip/CIP-0181))

Removes the requirement for a DRep delegation to be present when withdrawing staking rewards, reducing friction for ada holders who want to withdraw without participating in governance.

### Structural Groundwork for Phases 2 & 3

These are not feature activations. They are the block structure changes, header extensions, and protocol parameter introductions that Phases 2 and 3 depend on. Because all of these require a new era, they must ship in Phase 1. Any protocol parameters needed to govern Leios or Peras behaviour must also be defined here, even if their values are not yet active, since protocol parameters cannot be introduced in an intra-era hard fork.

| Change | Purpose |
|--------|---------|

> | [CIP-140](https://cips.cardano.org/cip/CIP-0140) Peras header extensions | Reserves space for Peras voting fields in block headers without activating the voting layer |

Reviewer 2: As far as I know, Peras doesn't change the header, so both columns of those row seem inaccurate to me. Peras votes and certs exist outside of the chain. Occasionally, when Peras is struggling, a block _body_ will contain a Peras cert. But as far as I know, the block header doesn't change at all for Peras.

> | [CIP-164](https://cips.cardano.org/cip/CIP-0164) Leios block header and body extensions | Introduces the endorser block structures, ranking block header extensions, and protocol parameters Linear Leios requires |

Reviewer 2: The left column correctly mentions "body extensions", but the right column doesn't mention the part of change to block bodies: that they can sometimes contain a Leios cert.

### Conditional: Fair Min Fees ([CIP-23](https://cips.cardano.org/cip/CIP-0023))

The protocol parameter introduced by CIP-23 is defined at Phase 1, but the fee rule itself may or may not activate at the Phase 1 hard fork. If implementation and governance readiness align, CIP-23 activates with Phase 1. If not, the parameter definition ships dormantly and the fee rule activates in a subsequent intra-era hard fork within the Dijkstra era.

---

## Phase 2: Linear Leios Activation (Intra-era Hard Fork, Q1 2027)

### What Changes

This phase activates Ouroboros Linear Leios via an intra-era hard fork, a protocol version bump within the Dijkstra era. The CIP-164 block body extensions shipped with Phase 1, so no new ledger era is needed. The structural groundwork is already in place; this hard fork activates the consensus rules that switch it on.

> Linear Leios is a partial realisation of the Leios throughput vision. It delivers meaningful gains but does not achieve everything Full Leios would. The Leios Innovation Team concluded that no safe design for Full Leios has been found: the closest approach required sharding the ledger, which dapp developers and users identified as highly undesirable. Rather than stall on an unsolved problem, the team chose to ship what works. Full Leios remains aspirational. Any path to it would require major changes to block structure and cryptography and a new ledger era, Euler or later.

Reviewer 2: This text is as blunt as my reply to your DM was, which seems maybe excessive in this context. I don't want to undersell what was studied for Full Leios.

But the gist is accurate: we didn't see a path for the first version of Leios on mainnet to have all the complexities (which weren't 100% pinned down) that Full Leios would require.

> block structure and cryptography and a new ledger era

This wording seems a bit off, in my interpretation. A couple clarifications:

- From what I recall, it wouldn't require _new_ crypto. Linear Leios already introduces votes, and I think Full Leios could use the same voting crypto---it's just the rules for when honest nodes should vote would change. And possibly also the bytes that they vote is signing. Also, Full Leios assumed a parallel leader schedule (for when to issue Input Blocks)---but again, I think that would just be another instances of the VRF crypto we already have.

- "block structure" would change, ~but it'd be the structure of the new Leios blocks~. The Praos blocks would be the same as in Linear Leios: they might contain a (Linear/Full) Leios cert---but the Linear->Full change might not even require a new codec. ... ah; I do recall that _tx_'s potentially would need some new fields, so that would indirectly be a block codec change.

(Maybe these distinctions are finer than this document's audience cares about?)

Linear Leios increases Cardano's throughput without changing the security guarantees of the base protocol. The original Leios research design used three block types including Input Blocks; Linear Leios eliminates those and works with two:

- **Ranking Blocks (RBs)** are the existing Praos blocks, extended with optional fields to announce and certify Endorser Blocks.
- **Endorser Blocks (EBs)** are larger supplementary blocks that contain references to additional transactions, not the transactions themselves.

> Transactions continue to propagate through the standard mempool. When a block producer wins slot leadership, it produces an RB that optionally announces an Endorser Block. The announcement is part of the RB itself and contains the EB, which holds the hashes of the transactions being endorsed. Nodes that do not already have those transactions request them from peers via the node-to-node protocol. A stake-based committee then certifies the EB; the certificate contains the hash of the EB and the aggregated signatures proving a 75% quorum of active stake. A subsequent RB includes that certificate, applying the endorsed transactions to the ledger. If no certified EB is available, a Ranking Block includes transactions directly as in standard Praos. The result is that the network can absorb significantly more transactions per unit of time without requiring larger blocks or faster slots.

Reviewer 2: ```suggestion
Transactions continue to propagate through the standard mempool. When a block producer wins slot leadership, it produces an RB that optionally announces an Endorser Block. The announcement is part of the RB itself and contains the EB, which holds the hashes of the transactions being endorsed. Nodes that do not already have those transactions request them from peers via some new node-to-node protocols. A stake-based committee then certifies the EB; the certificate contains the hash of the EB and the aggregated signatures proving a 75% quorum of active stake. A subsequent RB includes that certificate, applying the endorsed transactions to the ledger. If no certified EB is available, a Ranking Block includes transactions directly as in standard Praos. The result is that the network can absorb significantly more transactions per unit of time without requiring larger blocks or faster slots.
```

> Transactions continue to propagate through the standard mempool. When a block producer wins slot leadership, it produces an RB that optionally announces an Endorser Block. The announcement is part of the RB itself and contains the EB, which holds the hashes of the transactions being endorsed. Nodes that do not already have those transactions request them from peers via the node-to-node protocol. A stake-based committee then certifies the EB; the certificate contains the hash of the EB and the aggregated signatures proving a 75% quorum of active stake. A subsequent RB includes that certificate, applying the endorsed transactions to the ledger. If no certified EB is available, a Ranking Block includes transactions directly as in standard Praos. The result is that the network can absorb significantly more transactions per unit of time without requiring larger blocks or faster slots.

Reviewer 2: ```suggestion
Transactions continue to propagate through the standard mempool. When a block producer wins slot leadership, it produces an RB that optionally announces an Endorser Block. The announcement is part of the RB header itself and contains the hash of the EB, which in turn holds the hashes of the transactions being endorsed. Nodes that do not already have those transactions request them from peers via the node-to-node protocol. A stake-based committee then certifies the EB; the certificate contains the hash of the EB and the aggregated signatures proving a 75% quorum of active stake. A subsequent RB includes that certificate, applying the endorsed transactions to the ledger. If no certified EB is available, a Ranking Block includes transactions directly as in standard Praos. The result is that the network can absorb significantly more transactions per unit of time without requiring larger blocks or faster slots.
```

### Rollout Milestones

| Milestone | Notes |
|-----------|-------|
| Node release | Linear Leios-compatible node released for testnet operators |
| **Preview hard fork** | Governance action submitted and enacted on Preview; SPO testing window opens |

> | Preview SPO testing window | ~2 weeks; IB/EB propagation testing, throughput benchmarking |

Reviewer 2: ```suggestion
| Preview SPO testing window | ~2 weeks; EB propagation testing, throughput benchmarking |
```

| **Pre-production hard fork** | Governance action submitted and enacted on Pre-production; SPO testing window opens |
| Pre-production SPO testing window | ~1-2 weeks; final readiness checks |
| Mainnet governance action submission | Governance action submitted on Mainnet; DReps, SPOs, and Constitutional Committee voting period |
| **Mainnet activation** | Following governance ratification; date TBD |

### Activates

> - [CIP-164](https://cips.cardano.org/cip/CIP-0164): Ouroboros Linear Leios, full activation of input block production and endorser block certification

Reviewer 2: > _full_ activation

I worry about "full" here. I strongly doubt we should enable Leios on mainet at full strength immediately. The increase in throughput should be gradual (spread over multiple increases to the protocol parameters that limit the size of EBs and the cumulative txs they reference).

But maybe that's beyond the scope of this document because those parameters can be ratcheted via normal, non-HF governance actions?

- [CIP-23](https://cips.cardano.org/cip/CIP-0023): Fair Min Fees fee rule, if not activated in Phase 1

### Prerequisites

- Phase 1 mainnet live (CIP-164 block body extensions in place)
- Linear Leios node implementation merged to mainline
- Governance action ratified on-chain by DReps, SPOs, and Constitutional Committee

---

## Phase 3: Peras Activation (Intra-era Hard Fork, Q2 2027)

### What Changes

> This phase activates Ouroboros Peras via an intra-era hard fork, a protocol version bump within the Dijkstra era. The CIP-140 header extensions shipped with Phase 1, so no new ledger era is needed. The structural groundwork is already in place; this hard fork activates the consensus rules that switch it on.

Reviewer 2: ```suggestion
This phase activates Ouroboros Peras via an intra-era hard fork, a protocol version bump within the Dijkstra era. The CIP-140 codec extensions shipped with Phase 1, so no new ledger era is needed. The structural groundwork is already in place; this hard fork activates the consensus rules that switch it on.
```
